Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.ChamberMagazine.cs

```diff
@@ -0,0 +1,126 @@
+        return slot.ContainedEntity;
+    }
+
+    protected (int, int) GetChamberMagazineCountCapacity(ChamberMagazineAmmoProviderComponent component)
+    {
+        var count = GetChamberEntity(component.Owner) != null ? 1 : 0;
+        var (magCount, magCapacity) = GetMagazineCountCapacity(component);
+        return (count + magCount, magCapacity);
+    }
+
+    private bool TryInsertChamber(EntityUid uid, EntityUid ammo)
+    {
+        return Containers.TryGetContainer(uid, ChamberSlot, out var container) &&
+               container is ContainerSlot slot &&
+               slot.Insert(ammo);
+    }
+
+    private void OnChamberMagazineTakeAmmo(EntityUid uid, ChamberMagazineAmmoProviderComponent component, TakeAmmoEvent args)
+    {
+        // So chamber logic is kinda sussier than the others
+        // Essentially we want to treat the chamber as a potentially free slot and then the mag as the remaining slots
+        // i.e. if we shoot 3 times, then we use the chamber once (regardless if it's empty or not) and 2 from the mag
+        // We move the n + 1 shot into the chamber as we essentially treat it like a stack.
+        TryComp<AppearanceComponent>(uid, out var appearance);
+
+        if (TryTakeChamberEntity(uid, out var chamberEnt))
+        {
+            args.Ammo.Add(EnsureComp<AmmoComponent>(chamberEnt.Value));
+        }
+
+        var magEnt = GetMagazineEntity(uid);
+
+        // Pass an event to the magazine to get more (to refill chamber or for shooting).
+        if (magEnt != null)
+        {
+            // We pass in Shots not Shots - 1 as we'll take the last entity and move it into the chamber.
+            var relayedArgs = new TakeAmmoEvent(args.Shots, new List<IShootable>(), args.Coordinates, args.User);
+            RaiseLocalEvent(magEnt.Value, relayedArgs);
+
+            // Put in the nth slot back into the chamber
+            // Rest of the ammo gets shot
+            if (relayedArgs.Ammo.Count > 0)
+            {
+                var newChamberEnt = ((AmmoComponent) relayedArgs.Ammo[^1]).Owner;
+                TryInsertChamber(uid, newChamberEnt);
+            }
+
+            // Anything above the chamber-refill amount gets fired.
+            for (var i = 0; i < relayedArgs.Ammo.Count - 1; i++)
+            {
+                args.Ammo.Add(relayedArgs.Ammo[i]);
+            }
+        }
+        else
+        {
+            appearance?.SetData(AmmoVisuals.MagLoaded, false);
+            return;
+        }
+
+        var count = chamberEnt != null ? 1 : 0;
+        var capacity = 1;
+
+        if (TryComp<AppearanceComponent>(magEnt, out var magAppearance))
+        {
+            magAppearance.TryGetData<int>(AmmoVisuals.AmmoCount, out var addCount);
+            magAppearance.TryGetData<int>(AmmoVisuals.AmmoMax, out var addCapacity);
+            count += addCount;
+            capacity += addCapacity;
+        }
+
+        FinaliseMagazineTakeAmmo(uid, component, args, count, capacity, appearance);
```

Comment:
..why is there shared code that has actual logic dictated by stuff in appearance data? this seems weird

Reply:
lazy. there's an event for getting ammo count but I don't think I implemented it for every ammoprovider so I'll do that.